### PR TITLE
adding AVL tree

### PR DESCRIPTION
**Pull request by Zzocker**
- Added AVL tree
Signed-off-by: Zzocker <pkspritam16@gmail.com>

### CODE REVIEW

File: datastructures/avl/avl_test.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
+package avl
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (tjgurwara99) [inline datastructures/avl/avl_test.go:6]: Can you not do it using the standard library?

It seems the `testify` package has a lot of dependencies. I think its easily testable without this package.

Comment (Zzocker) [reply datastructures/avl/avl_test.go]: removed the ``testify`` dependencies.

File: datastructures/avl/avl.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	} else if bFactor == -2 { // R
+		bFactor = balanceFactor((*root).Right)
+		if bFactor == 1 { // RL
+			rlRotation(root)
+		} else if bFactor == -1 { // RR
+			rrRotation(root)
+		}
+	}
+}
+func delete(root **Node, key int) {
```
Comment (tjgurwara99) [inline datastructures/avl/avl.go:70]: Do you think this function should be exported?

**Review by tjgurwara99 — DISMISSED** [review]
Everything looks good to me!

**Review by iamrajiv — DISMISSED** [review]

**Review by raklaptudirm — APPROVED** [review]

File: datastructures/avl/avl.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	if root.Left != nil {
+		leftHeight = root.Left.Height
+	}
+	if root.Right != nil {
+		rightHeight = root.Right.Height
+	}
+	if rightHeight > leftHeight {
+		leftHeight = rightHeight
+	}
+	return 1 + leftHeight // 1+ max(leftHeight,rightHeight)
```
Comment (siriak) [inline datastructures/avl/avl.go:211]: Why not write as in the comment with max? Assigning rightHeight to leftHeight is really misleading

**Review by siriak — APPROVED** [review]
LGTM, thanks!

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**